stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -2,30 +2,233 @@
 
 package com.facebook.stetho.inspector.protocol.module;
 
+import com.facebook.stetho.common.LogUtil;
+import com.facebook.stetho.common.Util;
+import com.facebook.stetho.inspector.elements.DOMProvider;
+import com.facebook.stetho.inspector.elements.NodeAttribute;
+import com.facebook.stetho.inspector.elements.NodeDescriptor;
+import com.facebook.stetho.inspector.elements.NodeType;
+import com.facebook.stetho.inspector.helper.ChromePeerManager;
+import com.facebook.stetho.inspector.helper.ObjectIdMapper;
+import com.facebook.stetho.inspector.helper.PeersRegisteredListener;
 import com.facebook.stetho.inspector.jsonrpc.JsonRpcPeer;
 import com.facebook.stetho.inspector.jsonrpc.JsonRpcResult;
 import com.facebook.stetho.inspector.protocol.ChromeDevtoolsDomain;
 import com.facebook.stetho.inspector.protocol.ChromeDevtoolsMethod;
+import com.facebook.stetho.json.ObjectMapper;
 import com.facebook.stetho.json.annotation.JsonProperty;
 
 import org.json.JSONObject;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
 public class DOM implements ChromeDevtoolsDomain {
-  public DOM() {
+  private final ChromePeerManager mPeerManager;
+  private final DOMProvider.Factory mDOMProviderFactory;
+
+  private ObjectMapper mObjectMapper;
+  private DOMProvider mDOMProvider;
+  private DOMObjectIdMapper mObjectIdMapper;
+
+  public DOM(DOMProvider.Factory providerFactory) {
+    mDOMProviderFactory = Util.throwIfNull(providerFactory);
+
+    mPeerManager = new ChromePeerManager();
+    mPeerManager.setListener(new PeerManagerListener());
+  }
+
+  private final class PeerManagerListener extends PeersRegisteredListener {
+    @Override
+    protected void onFirstPeerRegistered() {
+      assert(mDOMProvider == null);
```

Comment:
Do not use `assert`.  It's tragically broken in Java.  Use `Util.throwIfNotNull`.  Also, you have a race here if you don't synchronize these two methods.  The base class doesn't do it because I was trying not to be heavy handed (maybe I should've though)